### ISSUE TITLE
e-sparqlEndpoint should support paged queries

### ISSUE THREAD

**Issue by jindrichmynarz** [UnifiedViews/Plugins#241]
e-sparqlEndpoint should support splitting queries via paging (LIMIT/OFFSET) to enable extraction of more RDF triples than the maximum number that is allowed per single request (e.g., via ResultSetMaxRows in virtuoso.ini). For example, the maximum number or triples per single request is set to 10000 by default in Virtuoso.

When you want to extract more triples, you to do one of two things. If you control the SPARQL endpoint you extract data from, then you can increase the setting for the maximum number of triples for a single request. If you don't control the SPARQL endpoint you extract data from, you need to split your query into multiple queries that retrieve parts of the complete results. Splitting can be done via paging by LIMIT and OFFSET. Since paging requires stable sort order, queries should be wrapped in subqueries to enable to use "scrollable cursors" (see mff-uk/DPUs#78).

A rough Clojure implementation of a SPARQL extractor supporting paged queries based on Apache Jena can be found at <https://github.com/opendatacz/sparql-unlimited>.

An alternative to the LIMIT/OFFSET approach can use the last seen ID (URI), as explained [here](http://use-the-index-luke.com/no-offset) (mentioned by @hanecak [here](https://team.eea.sk/wiki/pages/viewpage.action?pageId=110463376&focusedCommentId=118391471#comment-118391471)).

**Comment by mkl-eea-sk** [comment]
I do not see any DPU e-sparqlEndpoint. Where it is? Is it e-rdfFromSparql after rework (#178) ? If so, this task is blocked by #178

**Comment by jindrichmynarz** [comment]
It was in the (now deleted) branch dpu/e-sparqlEndpoint. It is also in https://github.com/mff-uk/DPUs/tree/master/dpu/e-sparqlEndpoint.

**Comment by mkl-eea-sk** [comment]
Pfuu, so lets clarify what is what and what has to be done. I am assigning this to @tomas-knap to clarify. This issue is blocked.

**Comment by skrchnavy** [comment]
see #222 for details why is branch missing

**Comment by mkl-eea-sk** [comment]
merge issues #241 and  #178

**Comment by tomas-knap** [comment]
This issue is blocked by #178, it is an extension of #178

**Comment by skrchnavy** [comment]
Will be handled when #266 closed in extra feature branch.

**Comment by tomas-knap** [comment]
@skrchnavy This will be solved in the separate branch, right? So it is not ready for test, I guess

**Comment by skrchnavy** [comment]
Yes, we will start when #266 finalised in develop.

**Comment by jindrichmynarz** [comment]
As discussed with @mkl-eea-sk today over Skype I'm adding more information. As I thought more about this issue, I realized it is actually simpler than implementing pages updates as I have done in [here](https://github.com/opendatacz/sparql-unlimited).

Let's start with example queries. If we want to retrieve all villages (instances of the class `dbo:Village`) from DBpedia (<http://dbpedia.org/sparql>), we can start with the following query:

```
PREFIX dbo:  <http://dbpedia.org/ontology/>

CONSTRUCT {
  ?village a dbo:Village .
}
WHERE {
  ?village a dbo:Village .
}
```

This will return 10001 triples, because the `ResultSetMaxRows` in virtuoso.ini is set to 10001 (or maybe is set to 10000 as is the default and Virtuoso strangely allows to retrieve one more triple). However, there are 155 597 villages in DBpedia as the following count query tells us:

```
PREFIX dbo:  <http://dbpedia.org/ontology/>

SELECT (COUNT(?village) AS ?countOfVillages)
WHERE {
  ?village a dbo:Village .
}
```

Since we want to retrieve all of them, we need to split our queries into pages sized as the maximum numbers of rows DBpedia allows to return. In order to split results into pages, we use `LIMIT` and `OFFSET`:

```
PREFIX dbo:  <http://dbpedia.org/ontology/>

CONSTRUCT {
  ?village a dbo:Village .
}
WHERE {
  ?village a dbo:Village .
}
LIMIT 10000
OFFSET 0
```

Here, `LIMIT` defines the page size. In general, it will be set by a user to a number lesser than the maximum number of results a SPARQL endpoint allows to return.

However, in order to maintain stable sort order when paging through the results, we need to explicitly sort the projected bindings. In our query, there is only one binding in the outermost WHERE clause, which is `?village`. Therefore, we can add an `ORDER BY` clause to our query (side note: default sort is ascending):

```
PREFIX dbo:  <http://dbpedia.org/ontology/>

CONSTRUCT {
  ?village a dbo:Village .
}
WHERE {
  ?village a dbo:Village .
}
ORDER BY ?village
LIMIT 10000
OFFSET 0
```

Unfortunately, when we iterate by adding `LIMIT` to `OFFSET` and reach the `OFFSET` of 40 000, DBpedia's SPARQL endpoint throws an error `Virtuoso 22023 Error SR353: Sorted TOP clause specifies more then 50000 rows to sort. Only 40000 are allowed. Either decrease the offset and/or row count or use a scrollable cursor`.

```
PREFIX dbo:  <http://dbpedia.org/ontology/>

CONSTRUCT {
  ?village a dbo:Village .
}
WHERE {
  ?village a dbo:Village .
}
ORDER BY ?village
LIMIT 10000
OFFSET 40000
```

This is caused by the `MaxSortedTopRows` from virtuoso.ini being set to 40 000. To avoid this problem, we need to use what Virtuoso calls a scrollable cursor, which means wrapping our query in a sub-SELECT (the comments describe how the original query is used):

```
PREFIX dbo:  <http://dbpedia.org/ontology/>

CONSTRUCT {
  ?village a dbo:Village .
}
WHERE {
  {
    SELECT ?village # All variables from the WHERE clause
    WHERE { # The original WHERE clause
      ?village a dbo:Village .
    }
    ORDER BY ?village # Sort by a selected variable
  }
}
LIMIT 10000
OFFSET 40000
```

Finally, we need to determine a stopping condition for iteration. The paging will stop when empty result set is retrieved.

Because of the possibility of such queries, it is important to use the bindings projected by the outermost WHERE clause and not the CONSTRUCT clause. If you are fortunate, a library may do this for you (i.e. its object representation will mirror the WHERE clause as the CONSTRUCT clause). Since there are 155 597 villages in DBpedia (as of July 22, 2015), a query with a greater `OFFSET` will produce empty result set:

```
PREFIX dbo:  <http://dbpedia.org/ontology/>

CONSTRUCT {
  ?village a dbo:Village .
}
WHERE {
  {
    SELECT ?village
    WHERE {
      ?village a dbo:Village .
    }
    ORDER BY ?village
  }
}
LIMIT 10000
OFFSET 160000
```

This query produces `# Empty TURTLE`. Fuseki, on the other hand returns the list of pre-defined prefixes:

```
@prefix dc:    <http://purl.org/dc/elements/1.1/> .
@prefix :      <http://example.org/book/> .
@prefix ns:    <http://example.org/ns#> .
@prefix dbo:   <http://dbpedia.org/ontology/> .
@prefix vcard: <http://www.w3.org/2001/vcard-rdf/3.0#> .
```

Both responses, when parsed as RDF, contain no triples. Therefore the stopping condition is discovering that a response contains 0 triples.

**Warning:** Note that our simple CONSTRUCT query can be shortened as the following:

```
PREFIX dbo:  <http://dbpedia.org/ontology/>

CONSTRUCT WHERE {
  ?village a dbo:Village .
}
```

Context: the primary RDF store to be supported is Virtuoso.

This is the core principle of query rewriting. As a follow-up I will prepare several queries and manually rewrite them.

**Comment by jindrichmynarz** [comment]
Here are examples of SPARQL queries manually transformed to use paging. All examples transform the input query into a paged query assuming the first page is requested and that the page size is 10 000. It seems to me that there will not be many different cases and so only a few tests can cover a lot.

These can be used to test query rewriting for paged query execution in e-sparqlEndpoint. However, note that this cannot be tested via string equality as several aspects of SPARQL queries are arbitrary; e.g., white-space and indentation, order of solution modifiers (such as `LIMIT` and `OFFSET`). Solutions including using a RDF library that provides SPARQL query equality tests or deterministic serialization of SPARQL queries. 

## 1. Basic query

### Input

```
PREFIX dbo:  <http://dbpedia.org/ontology/>

CONSTRUCT {
  ?village a dbo:Village .
}
WHERE {
  ?village a dbo:Village .
}
```

### Expected output

```
PREFIX dbo:  <http://dbpedia.org/ontology/>

CONSTRUCT {
  ?village a dbo:Village .
}
WHERE {
  {
    SELECT ?village
    WHERE {
      ?village a dbo:Village .
    }
    ORDER BY ?village
  }
}
LIMIT 10000
OFFSET 0
```

## 2. Shortened version of the basic query

### Input

```
PREFIX dbo:  <http://dbpedia.org/ontology/>

CONSTRUCT WHERE {
  ?village a dbo:Village .
}
```

### Expected output

```
PREFIX dbo:  <http://dbpedia.org/ontology/>

CONSTRUCT {
  ?village a dbo:Village .
}
WHERE {
  {
    SELECT ?village
    WHERE {
      ?village a dbo:Village .
    }
    ORDER BY ?village
  }
}
LIMIT 10000
OFFSET 0
```

## 3. Basic query with blank nodes

This demonstrates that we cannot use variables from the `CONSTRUCT` clause but instead we must extract them from the outermost `WHERE` clause.
 
### Input

```
PREFIX dbo:  <http://dbpedia.org/ontology/>

CONSTRUCT {
  _:village a dbo:Village .
}
WHERE {
  ?village a dbo:Village .
}
```

### Expected output

```
PREFIX dbo:  <http://dbpedia.org/ontology/>

CONSTRUCT {
  _:village a dbo:Village .
}
WHERE {
  {
    SELECT ?village
    WHERE {
      ?village a dbo:Village .
    }
    ORDER BY ?village
  }
}
LIMIT 10000
OFFSET 0
```

## 4. Villages located in more than 1 country

Note that aggregates (i.e. `GROUP BY`) are computed only after getting query solutions, so we need to wrap them in a sub-SELECT instead of directly using `GROUP BY` with `CONSTRUCT`.

### Input

```
PREFIX dbo: <http://dbpedia.org/ontology/>

CONSTRUCT {
  ?village a dbo:Village .
}
WHERE {
  {
    SELECT ?village
    WHERE {
      GRAPH <http://dbpedia.org> {
        ?village a dbo:Village ;
          dbo:country ?country .
      }
    }
    GROUP BY ?village
    HAVING (COUNT(DISTINCT ?country) > 1)
  }
}
```

### Expected output

```
PREFIX dbo: <http://dbpedia.org/ontology/>

CONSTRUCT {
  ?village a dbo:Village .
}
WHERE {
  {
    SELECT ?village
    WHERE {
      {
        SELECT ?village
        WHERE {
          GRAPH <http://dbpedia.org> {
            ?village a dbo:Village ;
              dbo:country ?country .
          }
        }
        GROUP BY ?village
        HAVING (COUNT(DISTINCT ?country) > 1)
      }
    }
    ORDER BY ?village
  }
}
LIMIT 10000
OFFSET 0
```

**Comment by mkl-eea-sk** [comment]
## Query A. (from 4. Villages located in more than 1 country)
Anonymous _:village in construct projection.

Note that aggregates (i.e. `GROUP BY`) are computed only after getting query solutions, so we need to wrap them in a sub-SELECT instead of directly using `GROUP BY` with `CONSTRUCT`.

### Input

```
PREFIX dbo: <http://dbpedia.org/ontology/>

CONSTRUCT {
  _:village a dbo:Village .
}
WHERE {
  {
    SELECT ?village
    WHERE {
      GRAPH <http://dbpedia.org> {
        ?village a dbo:Village ;
          dbo:country ?country .
      }
    }
    GROUP BY ?village
    HAVING (COUNT(DISTINCT ?country) > 1)
  }
}
```

### Expected output

```
PREFIX dbo: <http://dbpedia.org/ontology/>

CONSTRUCT {
  ?village a dbo:Village .
}
WHERE {
  {
    SELECT ?village
    WHERE {
      {
        SELECT ?village
        WHERE {
          GRAPH <http://dbpedia.org> {
            ?village a dbo:Village ;
              dbo:country ?country .
          }
        }
        GROUP BY ?village
        HAVING (COUNT(DISTINCT ?country) > 1)
      }
    }
    ORDER BY ?village
  }
}
LIMIT 10000
OFFSET 0
```

**Comment by mkl-eea-sk** [comment]
## Query B (from 5. Names of villages located in more than 1 country)
anon projections in construct
### Input

```
PREFIX dbo:  <http://dbpedia.org/ontology/>
PREFIX rdfs:  <http://www.w3.org/2000/01/rdf-schema#>

CONSTRUCT {
  _:village a dbo:Village ;
    rdfs:label _:label .
}
WHERE {
  {
    SELECT ?village
    WHERE {
      GRAPH <http://dbpedia.org> {
        ?village a dbo:Village ;
          dbo:country ?country .
      }
    }
    GROUP BY ?village
    HAVING (COUNT(DISTINCT ?country) > 1)
  }
  GRAPH <http://dbpedia.org> {
    ?village rdfs:label ?label .
  }
}
```

### Expected output

```
PREFIX dbo:  <http://dbpedia.org/ontology/>
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>

CONSTRUCT {
  ?village a dbo:Village ;
    rdfs:label ?label .
}
WHERE {
  {
    SELECT ?village ?label
    WHERE {
      {
        SELECT ?village
        WHERE {
          GRAPH <http://dbpedia.org> {
            ?village a dbo:Village ;
              dbo:country ?country .
          }
        }
        GROUP BY ?village
        HAVING (COUNT(DISTINCT ?country) > 1)
      }
      GRAPH <http://dbpedia.org> {
        ?village rdfs:label ?label .
      }
    }
    ORDER BY ?village
  }
}
LIMIT 10000
OFFSET 0
```

**Comment by jindrichmynarz** [comment]
The queries with blank nodes in `CONSTRUCT` clause likely indicate an error on the side of the query creator. Their semantics is different to what the query creator probably expects. Instead of binding projected variables to blank nodes (e.g., `_:village`), a new blank node is created for each binding. For example, if we have this query:

```
PREFIX dbo:  <http://dbpedia.org/ontology/>
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>

CONSTRUCT {
  _:village a dbo:Village ;
    rdfs:label _:label .
}
WHERE {
  ?village a dbo:Village ;
    rdfs:label ?label .
}
```

And we have this sample dataset:

```
@prefix dbo:  <http://dbpedia.org/ontology/> .
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .

<http://dbpedia.org/resource/Jeenmata> a dbo:Village ;
   rdfs:label "Jeenmata" .

<http://dbpedia.org/resource/Vallangi> a dbo:Village ;
   rdfs:label "Vallangi" .
```

The result is:

```
@prefix rdfs:    <http://www.w3.org/2000/01/rdf-schema#> .
@prefix dbo:     <http://dbpedia.org/ontology/> .

[]    a dbo:Village ;
      rdfs:label    [] .

[]    dbo:Village ;
      rdfs:label    [] .
```

That is, URIs and labels of villages are lost and replaced by fresh blank nodes. The number of villages in the result is the same as in the output.

**Comment by tomas-knap** [comment]
@mkl-eea-sk So what is the current status? What is planned to be done?

**Comment by mkl-eea-sk** [comment]
Query rewriting as described here in in implementation phase.

**Comment by tomas-knap** [comment]
@mkl-eea-sk  Ok, yes, but what is exactly being done? Support for rewritting the five types of queries as introduced above? 

@jindrichmynarz Such extension as proposed above is usefull only for Virtuoso SPARQL endpoints, right?

**Comment by mkl-eea-sk** [comment]
@tomas-knap those are test cases for junits

**Comment by jindrichmynarz** [comment]
@tomas-knap: Since e-sparqlEndpoint supports only `CONSTRUCT`, query rewriting is being implemented only for this query type.

Paged query execution is useful for any RDF store that can be configured to limit resources spent on query execution (e.g., maximum response size, time-out). In general, every sane maintainer of a public SPARQL endpoint should use some of these settings. For example, in case of Virtuoso, it can be done via `ResultSetMaxRows`, while Fuseki has `arq:queryTimeout` that limits maximum query execution time. Therefore, splitting a more demanding query into several less demanding queries that can be executed within the limits imposed by SPARQL endpoint is generally useful.

However, the part with nested sub-SELECT for the "scrollable cursor" is indeed Virtuoso-specific. Nevertheless, the rewritten query remains valid SPARQL and can be used in other RDF stores without a problem.

**Comment by mkl-eea-sk** [comment]
@jindrichmynarz at query5, please, will it make any difference if we order by village or by label?

**Comment by jindrichmynarz** [comment]
It should not make a difference. But that is what we have to test.

Basically, we have 2 options:

1. ORDER BY an arbitrary projected variable (e.g., `ORDER BY ?village`)
2. ORDER BY all projected variables (e.g., `ORDER BY ?village ?label`)

We need to test whether option 1 produces all results and the pages are not overlapping. If not, we need to use option 2.

**Comment by mkl-eea-sk** [comment]
Ok, currently my code produces ORDER BY ?label :)

**Comment by mkl-eea-sk** [comment]
And it is even easier to go for option 2

**Comment by jindrichmynarz** [comment]
OK, let's go straight for option. Still we should test this approach is valid. I'm thinking about preparing several sample query results in RDF that we can load in an RDF store, query with the paged execution and the compare if the results are the same as the input samples.

**Comment by tomas-knap** [comment]
Btw. at the end, please, provide possibility to be able to turn on/off the paged queries. By default, it should be turned on. This option can be put below the query.

**Comment by tomas-knap** [comment]
One follow-up issue: https://github.com/UnifiedViews/Plugins/issues/328
